File: Src/ButterflyNet.Parser/Strategies/TableStrategy.cs
using System;
using ButterflyNet.Parser.Satisfiers;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	public class TableStrategy : ScopeDrivenStrategy, ITokenProvider {
		public TableStrategy() {
			AddSatisfier<TableSatisfier>();
		}

		protected override void Execute(ParseContext context) {
			IScope rowScope;
			if (context.Input.Peek() == '{') {
				context.AdvanceInput();
				rowScope = new TableRowScope();
			} else {
				rowScope = new TableRowLineScope();
			}

			Type cellType = ScopeTypeCache.TableCell;
			if (context.Input.Peek() == '!') {
				context.AdvanceInput();
				cellType = ScopeTypeCache.TableHeader;
			}

			context.Input.SeekToNonWhitespace();
			context.UpdateCurrentChar();

			if (!context.Scopes.ContainsType(ScopeTypeCache.Table)) {
				//no tables, so create a new one
				OpenScope(new TableScope(), context);
				OpenScope(rowScope, context);
			} else if (context.Scopes.ContainsType(ScopeTypeCache.TableCell, ScopeTypeCache.TableHeader)) {
				//close table cell
				CloseScopeUntil(context, ScopeTypeCache.TableCell, ScopeTypeCache.TableHeader);
				CloseCurrentScope(context);
			} else if (!context.Scopes.ContainsType(ScopeTypeCache.TableRow, ScopeTypeCache.TableRowLine)) {
				OpenScope(rowScope, context);
			}

			//figure out if we need to open a new cell or not
			//criteria:
			// - not EOL
			// - we're in a row terminated by a line break and the next char is not a new line (signifying a new row)
			// - OR we're in a row not terminated by a line break
			if (
				context.Input.Peek() != ButterflyStringReader.NoValue 
					&& (
						(context.Scopes.ContainsType(ScopeTypeCache.TableRowLine) && context.Input.Peek() != '\n') 
							|| context.Scopes.ContainsType(ScopeTypeCache.TableRow)
						)
				) {
				OpenScope(cellType == ScopeTypeCache.TableHeader ? (IScope)new TableHeaderScope() : new TableCellScope(), context);
			}
		}

		private class TableSatisfier : ISatisfier {
			private readonly ISatisfier startOfLine = new StartOfLineSatisfier();
			private readonly ISatisfier containsTableScope = new InScopeStackSatisfier(ScopeTypeCache.Table);

			public bool IsSatisfiedBy(ParseContext context) {
				return startOfLine.IsSatisfiedBy(context) || containsTableScope.IsSatisfiedBy(context);
			}
		}

		public string Token { get { return "|"; } }
	}
}

File: src/js/src/strategies/TableStrategy.js
function TableStrategy() {
	TableStrategy.$parent.call(this);

	this.addSatisfier(new CannotNestInsideInlineSatisfier());
	this.addSatisfier(function() {
		var sol = new StartOfLineSatisfier();
		var containsTable = new InScopeStackSatisfier(ScopeTypeCache.table);
		
		return function(context) {
			return sol.isSatisfiedBy(context) || containsTable.isSatisfiedBy(context);
		};
	}());
	this.setAsTokenTransformer("|");
	
	this.doExecute = function(context) {
		var rowScope = new TableRowLineScope();
		if (context.input.peek() === '{') {
			context.advanceInput();
			rowScope = new TableRowScope();
		}

		var cellType = ScopeTypeCache.tableCell;
		if (context.input.peek() === '!') {
			context.advanceInput();
			cellType = ScopeTypeCache.tableHeader;
		}
		
		context.input.seekToNonWhitespace();
		context.updateCurrentChar();

		if (!context.scopes.containsType(ScopeTypeCache.table)) {
			//no tables, so create a new one
			this.openScope(new TableScope(), context);
			this.openScope(rowScope, context);
		} else if (context.scopes.containsType([ScopeTypeCache.tableCell, ScopeTypeCache.tableHeader])) {
			//close table cell
			var currentScope = context.scopes.peek();
			if (!currentScope || (currentScope.type !== ScopeTypeCache.tableCell && currentScope.type !== ScopeTypeCache.tableHeader)) {
				throw new ParseException("Cannot close table cell until all containing scopes are closed");
			}

			this.closeCurrentScope(context);
		} else if (!context.scopes.containsType([ScopeTypeCache.tableRow, ScopeTypeCache.tableRowLine])) {
			this.openScope(rowScope, context);
		}
		
		if (
			context.input.peek() !== ButterflyStringReader.EOF && (
				(context.scopes.containsType(ScopeTypeCache.tableRowLine) && context.input.peek() !== "\n") 
				|| context.scopes.containsType(ScopeTypeCache.tableRow)
			)
		) {
			this.openScope(cellType == ScopeTypeCache.tableHeader ? new TableHeaderScope() : new TableCellScope(), context);
		}
	};
}

extend(ScopeDrivenStrategy, TableStrategy);

function CloseTableRowStrategy() {
	CloseTableRowStrategy.$parent.call(this);
	
	this.addSatisfier(new InScopeStackSatisfier(ScopeTypeCache.tableRow));
	this.setAsTokenTransformer("}|");
	
	var closableScopes = [ScopeTypeCache.tableCell, ScopeTypeCache.tableHeader, ScopeTypeCache.paragraph];
	
	this.doExecute = function(context) {
		while (!context.scopes.isEmpty()) {
			if (!contains(closableScopes, context.scopes.peek().type)) {
				break;
			}

			this.closeCurrentScope(context);
		}

		var currentScope = context.scopes.peek();
		if (currentScope == null || currentScope.type !== ScopeTypeCache.tableRow) {
			throw new ParseException("Cannot close table row until all containing scopes are closed");
		}

		this.closeCurrentScope(context);
	};
}

extend(ScopeDrivenStrategy, CloseTableRowStrategy);